parteh/PRTAllometricCarbonMod.F90

```diff
@@ -0,0 +1,981 @@
+       c_mask(store_c_id)  = grow_store
+       c_mask(struct_c_id) = .true.                ! Always increment dead on growth step
+       c_mask(repro_c_id)  = .true.                ! Always calculate reproduction on growth
+       c_mask(dbh_id)      = .true.                ! Always increment dbh on growth step
+       
+       if(ODESolve == 2) then
+          this%ode_opt_step = totalC
+       end if
+       
+       do while( ierr .ne. 0 )
+          
+          deltaC = min(totalC,this%ode_opt_step)
+          if(ODESolve == 1) then
+             call RKF45(AllomCGrowthDeriv,c_pool,c_mask,deltaC,totalC, &
+                   max_trunc_error,intgr_params,c_pool_out,this%ode_opt_step,step_pass)
+             
+          elseif(ODESolve == 2) then
+             call Euler(AllomCGrowthDeriv,c_pool,c_mask,deltaC,totalC,intgr_params,c_pool_out)
+             !  step_pass = .true.
+             call CheckIntegratedAllometries(c_pool_out(dbh_id),ipft,canopy_trim,  &
+                   c_pool_out(leaf_c_id), c_pool_out(fnrt_c_id), c_pool_out(sapw_c_id), &
+                   c_pool_out(store_c_id), c_pool_out(struct_c_id), &
+                   c_mask(leaf_c_id), c_mask(fnrt_c_id), c_mask(sapw_c_id), &
+                   c_mask(store_c_id),c_mask(struct_c_id),  max_trunc_error, step_pass)
+             if(step_pass)  then
+                this%ode_opt_step = deltaC
+             else
+                this%ode_opt_step = 0.5*deltaC
+             end if
+          else
+             write(fates_log(),*) 'An integrator was chosen that DNE'
+             write(fates_log(),*) 'ODESolve = ',ODESolve
+             call endrun(msg=errMsg(sourcefile, __LINE__))
+          end if
+
+          nsteps = nsteps + 1
+          
+          if (step_pass) then ! If true, then step is accepted
+             totalC    = totalC - deltaC
+             c_pool(:) = c_pool_out(:)
+          end if
+          
+          if(nsteps > max_substeps ) then
+             write(fates_log(),*) 'Plant Growth Integrator could not find'
+             write(fates_log(),*) 'a solution in less than ',max_substeps,' tries'
+             write(fates_log(),*) 'Aborting'
+             write(fates_log(),*) 'carbon_balance',carbon_balance
+             write(fates_log(),*) 'deltaC',deltaC
+             write(fates_log(),*) 'totalC',totalC
+             write(fates_log(),*) 'leaf:',grow_leaf,target_leaf_c,target_leaf_c - leaf_c
+             write(fates_log(),*) 'fnrt:',grow_fnrt,target_fnrt_c,target_fnrt_c - fnrt_c
+             write(fates_log(),*) 'sap:',grow_sapw,target_sapw_c, target_sapw_c - sapw_c
+             write(fates_log(),*) 'store:',grow_store,target_store_c,target_store_c - store_c
+             write(fates_log(),*) 'dead:',target_struct_c,target_struct_c - struct_c
+             call endrun(msg=errMsg(sourcefile, __LINE__))
+          end if
+
+          ! TotalC should eventually be whittled down to near zero
+          ! At that point, update the actual states
```

Comment:
This is to account for slight numerical inconsistencies in the solver, right? (and to partition the C accordingly...)

Reply:
yes, updated comments